File: 14.december.2024/1.c
struct ListNode {
    int val;
    struct ListNode *next;
};




struct ListNode* addTwoNumbers(struct ListNode* l1,
                               struct ListNode* l2)
{
        struct ListNode start = {
                .val = -1;
                .next = NULL;
        };
        struct ListNode* sum = &start;
        struct ListNode* head = sum;

        int carry = 0;
        /*
         * While either of the list is not exhausted: add l1.val and l2.val to
         * sum.val and if sum.val > 10 then carry is sum.val%10.
         *
         */
        while ( l1 != NULL || l2  != NULL  ) {
                sum->val = l1->val + l2->val + carry;
                if (sum->val > 9) {
                        carry = (sum->val - sum->val%10)/10; /* Tenths place */
                        sum->val = sum->val%10; /* Units place */
                }

                struct ListNode next_number = {
                        .val = -1
                        .next = NULL
                };

                sum -> next = &next_number;
                sum = &next_number;
                l1 = l1 -> next;
                l2 = l2 -> next;
        }

        /* Which one did we exhaust; get the remaining elemens of the other one */
        if (l1 == NULL) {
                while (l2!=NULL) {
                        sum->val = l2->val + carry;
                        if (sum->val > 9) {
                                carry = (sum->val - sum->val%10)/10; /* Tenths place */
                                sum->val = sum->val%10; /* Units place */
                        }

                        struct ListNode next_number = {
                                .val = -1
                                .next = NULL
                        };

                        sum -> next = &next_number;
                        sum = &next_number;
                        l2 = l2 -> next;
                }
        }
        else {
                while (l1!=NULL) {
                        sum->val = l1->val + carry;
                        if (sum->val > 9) {
                                carry = (sum->val - sum->val%10)/10; /* Tenths place */
                                sum->val = sum->val%10; /* Units place */
                        }

                        struct ListNode next_number = {
                                .val = -1
                                .next = NULL
                        };

                        sum -> next = &next_number;
                        sum = &next_number;
                        l1 = l1 -> next;
                }
        }

        sum = NULL;

        return result;
}

File: 14.december.2024/1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self,
                      l1: Optional[ListNode],
                      l2: Optional[ListNode])
                        -> Optional[ListNode]:

        """
        We have 2 lists, and we need to add their corresponding
        elements. For example, we add 1st element of list 1 to
        first element of list 2 and so on..

        Now, we have 2 problems:

        1. What if the number we got by adding corresponding
        elements of list 1 and list 2 is more then 9. i.e. 2 digit.

        -> In this case, we store the unit's digit into the list
        and the 10th place is stored on carry. We add this carry
        to the sum of next 2 elements.

        e.g. 6 + 4 = 10. Hence carry = 1 and unit_place = 0. Hence
        0 goes into the current place and 1 becomes carry, which is
        added to next pair, i.e. 3 + 4 = 7. 7 + 1 (carry) = 8.
        8 goes into it.

        2. What if the lengths of 2 lists are not equal. For e.g.
        in [9,9,9,9,9,9,9], l2 = [9,9,9,9]

        Here we take the longer list and keep on repeting the
        process until the shorted list is exhausted. After that,
        we only add the carry and element of longer list. We
        would need to add another element, in case we have an
        outstanding carry at the end.
        """

        l = l1 if l1.len() > l2.len else l2"

        print(l1[0].val)
